# Housing Rental Analysis for San Francisco

In this challenge, your job is to use your data visualization skills, including aggregation, interactive visualizations, and geospatial analysis, to find properties in the San Francisco market that are viable investment opportunities.

## Instructions

Use the `san_francisco_housing.ipynb` notebook to visualize and analyze the real-estate data.

Note that this assignment requires you to create a visualization by using hvPlot and GeoViews. Additionally, you need to read the `sfo_neighborhoods_census_data.csv` file from the `Resources` folder into the notebook and create the DataFrame that you’ll use in the analysis.

The main task in this Challenge is to visualize and analyze the real-estate data in your Jupyter notebook. Use the `san_francisco_housing.ipynb` notebook to complete the following tasks:

* Calculate and plot the housing units per year.

* Calculate and plot the average prices per square foot.

* Compare the average prices by neighborhood.

* Build an interactive neighborhood map.

* Compose your data story.

### Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?

### Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

### Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

### Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `geo` parameter to True.
    * Set the `size` parameter to “sale_price_sqr_foot”.
    * Set the `color` parameter to “gross_rent”.
    * Set the `frame_width` parameter to 700.
    * Set the `frame_height` parameter to 500.
    * Include a descriptive title.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

### Compose Your Data Story

Based on the visualizations that you created, answer the following questions:

* How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

* What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

In [85]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path

## Import the data 

In [87]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_path= Path("./Resources/sfo_neighborhoods_census_data.csv")
sfo_data_df = pd.read_csv(sfo_data_path)
# Review the first and last five rows of the DataFrame
sfo_data_df.head()
sfo_data_df.tail()

year            neighborhood  sale_price_sqr_foot  housing_units  \
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390

---

## Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?



### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [89]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = sfo_data_df.groupby('year').mean()

# Review the DataFrame
housing_units_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualization.

In [90]:
# Create a visual aggregation explore the housing units by year
housing_units_by_year.hvplot.bar(x='year', y= "housing_units", title = "Housing Units 2010-2016")

:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analyzed?

**Answer:** # The average number of housing units in each neighborhood increases every year from 2010 - 2016.

---

## Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?



### Step 1: Group the data by year, and then average the results.

In [91]:
# Create a numerical aggregation by grouping the data by year and averaging the results
housing_units_by_year = sfo_data_df.groupby('year').mean()

# Review the resulting DataFrame
housing_units_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

**Answer:** The lowest gross rent reported is 1239 in 2010.

### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [93]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year=sfo_data_df.drop('housing_units',axis=1)
prices_square_foot_by_year.mean()
prices_square_foot_by_year


/Users/meghancolip/opt/anaconda3/envs/dev/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  after removing the cwd from sys.path.


year                   neighborhood  sale_price_sqr_foot  gross_rent
0    2010                   Alamo Square           291.182945        1239
1    2010                     Anza Vista           267.932583        1239
2    2010                        Bayview           170.098665        1239
3    2010               Buena Vista Park           347.394919        1239
4    2010               Central Richmond           319.027623        1239
5    2010                 Central Sunset           418.172493        1239
6    2010                 Corona Heights           369.359338        1239
7    2010                     Cow Hollow           569.379968        1239
8    2010                  Croker Amazon           165.645730        1239
9    2010                Diamond Heights           456.930822        1239
10   2010                      Downtown            241.804552        1239
11   2010  Eureka Valley/Dolores Heights           403.118313        1239
12   2010                      Excelsior           364.928336        1239
13   2010       Financial District North           401.749967        1239
14   2010       Financial District South           162.821210        1239
15   2010                  Forest Knolls           209.049327        1239
16   2010                      Glen Park           715.586438        1239
17   2010            Golden Gate Heights           551.801119        1239
18   2010                 Haight Ashbury           254.473779        1239
19   2010                   Hayes Valley           272.284980        1239
20   2010                  Hunters Point           170.624920        1239
21   2010                     Ingleside            275.056686        1239
22   2010                  Inner Mission           238.350020        1239
23   2010                 Inner Parkside           199.702004        1239
24   2010                 Inner Richmond           285.822358        1239
25   2010                   Inner Sunset           275.769730        1239
26   2010     Jordan Park/Laurel Heights           419.887812        1239
27   2010            Lake --The Presidio           320.796910        1239
28   2010                  Lone Mountain           398.458876        1239
29   2010          Lower Pacific Heights           374.057636        1239
30   2010                         Marina           387.392839        1239
31   2010                  Miraloma Park           680.608729        1239
32   2010                    Mission Bay           489.805378        1239
33   2010                Mission Dolores           257.987999        1239
34   2010                Mission Terrace           262.876026        1239
35   2010                       Nob Hill           307.677477        1239
36   2010                     Noe Valley           381.138445        1239
37   2010                      Oceanview           282.119671        1239
38   2010                 Outer Parkside           432.195292        1239
39   2010                Outer Richmond            338.126200        1239
40   2010                   Outer Sunset           368.311552        1239
41   2010                Pacific Heights           496.516014        1239
42   2010                     Park North           326.017150        1239
43   2010                       Parkside           274.166742        1239
44   2010      Parnassus/Ashbury Heights           434.255004        1239
45   2010                        Portola           302.156113        1239
46   2010                   Potrero Hill           491.450004        1239
47   2010               Presidio Heights           549.417931        1239
48   2010                   Russian Hill           362.348224        1239
49   2010                    South Beach          1037.099789        1239
50   2010                South of Market           401.295976        1239
51   2010                      Sunnyside           542.531682        1239
52   2010                 Telegraph Hill           524.793509        1239
53   2010

### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

> **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualization.


In [9]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title.
prices_square_foot_by_year.hvplot(x='year',ylabel="Price ($)", title ="Price of Unit per Year")

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** # Yes, there was a price drop during 2011.

**Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** # Gross rent increased in 2011.

---

## Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 


### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

In [94]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(['year','neighborhood']).mean()
prices_by_year_by_neighborhood

sale_price_sqr_foot  housing_units  \
year neighborhood                                                        
2010 Alamo Square                            291.182945       372560.0   
     Anza Vista                              267.932583       372560.0   
     Bayview                                 170.098665       372560.0   
     Buena Vista Park                        347.394919       372560.0   
     Central Richmond                        319.027623       372560.0   
     Central Sunset                          418.172493       372560.0   
     Corona Heights                          369.359338       372560.0   
     Cow Hollow                              569.379968       372560.0   
     Croker Amazon                           165.645730       372560.0   
     Diamond Heights                         456.930822       372560.0   
     Downtown                                241.804552       372560.0   
     Eureka Valley/Dolores Heights           403.118313       372560.0   
     Excelsior                               364.928336       372560.0   
     Financial District North                401.749967       372560.0   
     Financial District South                162.821210       372560.0   
     Forest Knolls                           209.049327       372560.0   
     Glen Park                               715.586438       372560.0   
     Golden Gate Heights                     551.801119       372560.0   
     Haight Ashbury                          254.473779       372560.0   
     Hayes Valley                            272.284980       372560.0   
     Hunters Point                           170.624920       372560.0   
     Ingleside                               275.056686       372560.0   
     Inner Mission                           238.350020       372560.0   
     Inner Parkside                          199.702004       372560.0   
     Inner Richmond                          285.822358       372560.0   
     Inner Sunset                            275.769730       372560.0   
     Jordan Park/Laurel Heights              419.887812       372560.0   
     Lake --The Presidio                     320.796910       372560.0   
     Lone Mountain                           398.458876       372560.0   
     Lower Pacific Heights                   374.057636       372560.0   
     Marina                                  387.392839       372560.0   
     Miraloma Park                           680.608729       372560.0   
     Mission Bay                             489.805378       372560.0   
     Mission Dolores                         257.987999       372560.0   
     Mission Terrace                         262.876026       372560.0   
     Nob Hill                                307.677477       372560.0   
     Noe Valley                              381.138445       372560.0   
     Oceanview                               282.119671       372560.0   
     Outer Parkside                          432.195292       372560.0   
     Outer Richmond                          338.126200       372560.0   
     Outer Sunset                            368.311552       372560.0   
     Pacific Heights                         496.516014       372560.0   
     Park North                              326.017150       372560.0   
     Parkside                                274.166742       372560.0   
     Parnassus/Ashbury Heights               434.255004       372560.0   
     Portola                                 302.156113       372560.0   
     Potrero Hill                            491.450004       372560.0   
     Presidio Heights                        549.417931       372560.0   
     Russian Hill                            362.348224       372560.0   
     South Beach                            1037.099789       372560.0   
     South of Market                         401.295976       372560.0   
     Sunnyside                               542.531682       372560.0   
     Telegraph Hill                     

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [12]:
# Filter out the housing_units
prices_by_year_by_neighborhood_no_housing_units = prices_by_year_by_neighborhood.drop('housing_units',axis=1)
# Review the first and last five rows of the DataFrame
prices_by_year_by_neighborhood_no_housing_units.head()
prices_by_year_by_neighborhood_no_housing_units.tail()

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

### Step 3: Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualization.

In [95]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
prices_by_year_by_neighborhood_no_housing_units.hvplot(x='year', ylabel='Price($)', groupby='neighborhood')

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualization to answer the following question:

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** The average sale price per square foot was more in 2016 than in 2012.

---

## Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `size_max` parameter to “25”.

    * Set the `zoom` parameter to “11”.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?


### Step 1: Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

In [97]:
# Load neighborhoods coordinates data
neighborhood_locations_df=pd.read_csv(
    Path("./Resources/neighborhoods_coordinates.csv"),
    index_col='Neighborhood',)





# Review the DataFrame
neighborhood_locations_df


Lat         Lon
Neighborhood                                        
Alamo Square                   37.791012 -122.402100
Anza Vista                     37.779598 -122.443451
Bayview                        37.734670 -122.401060
Bayview Heights                37.728740 -122.410980
Bernal Heights                 37.728630 -122.443050
Buena Vista Park               37.768160 -122.439330
Central Richmond               37.777890 -122.445170
Central Sunset                 37.749610 -122.489990
Clarendon Heights              37.753310 -122.447030
Corona Heights                 37.785530 -122.456000
Cow Hollow                     37.792980 -122.435790
Croker Amazon                  37.722800 -122.438690
Diamond Heights                37.728630 -122.443050
Downtown                       37.791010 -122.402100
Duboce Triangle                37.769640 -122.426110
Eureka Valley/Dolores Heights  37.755540 -122.437790
Excelsior                      37.728740 -122.410980
Financial District North       37.791010 -122.402100
Financial District South       37.791010 -122.402100
Forest Knolls                  37.757060 -122.455090
Glen Park                      37.728630 -122.443050
Golden Gate Heights            37.757912 -122.464020
Haight Ashbury                 37.765250 -122.435960
Hayes Valley                   37.765250 -122.435960
Hunters Point                  37.725510 -122.371780
Ingleside                      37.721161 -122.487068
Ingleside Heights              37.721160 -122.487070
Inner Mission                  37.754160 -122.419470
Inner Parkside                 37.749610 -122.489990
Inner Richmond                 37.769112 -122.483566
Inner Sunset                   37.749610 -122.489970
Jordan Park/Laurel Heights     37.785530 -122.456000
Lake --The Presidio            37.801520 -122.454570
Lone Mountain                  37.778900 -122.454960
Lower Pacific Heights          37.792980 -122.435790
Marina                         37.799720 -122.466890
Merced Heights                 37.719930 -122.465950
Midtown Terrace                37.753310 -122.447030
Miraloma Park                  37.734150 -122.457000
Mission Bay                    37.783230 -122.400650
Mission Dolores                37.751840 -122.425220
Mission Terrace                37.728630 -122.443050
Nob Hill                       37.792980 -122.435790
Noe Valley                     37.753310 -122.447030
North Beach                    37.800550 -122.404330
North Waterfront               37.800550 -122.404330
Oceanview                      37.719930 -122.465950
Outer Mission                  37.722800 -122.438690
Outer Parkside                 37.751700 -122.446000
Outer Richmond                 37.769110 -122.483570
Outer Sunset                   37.749610 -122.489990
Pacific Heights                37.792980 -122.435790
Park North                     37.728380 -122.478510
Parkside                       37.749610 -122.489990
Parnassus/Ashbury Heights      37.763550 -122.457650
Portola                        37.734670 -122.401060
Potrero Hill                   37.731960 -122.383240
Presidio Heights               37.785530 -122.456000
Russian Hill                   37.792980 -122.435790
Silver Terrace                 37.734670 -122.401060
South Beach                    37.783230 -122.400650
South of Market                37.791010 -122.402100
Sunnyside                      37.734150 -122.457000
Telegraph Hill                 37.800550 -122.404330
Twin Peaks                     37.753311 -122.447029
Union Square District          37.791010 -122.402100
Van Ness/ Civic Center         37.779510 -122.420220
Visitacion Valley              37.728740 -122.410980
West Portal                    37.740260 -122.463880
Western Addition               37.792980 -122.435790
Westwood Highlands             37.734700 -122.456854
Westwood Park                  37.734150 -122.457000
Yerba Buena                    37.792980 -122.396360

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

In [99]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo_data_df.groupby('neighborhood').mean()

# Review the resulting DataFrame
all_neighborhood_info_df

year  sale_price_sqr_foot  \
neighborhood                                                      
Alamo Square                   2013.000000           366.020712   
Anza Vista                     2013.333333           373.382198   
Bayview                        2012.000000           204.588623   
Bayview Heights                2015.000000           590.792839   
Bernal Heights                 2013.500000           576.746488   
Buena Vista Park               2012.833333           452.680591   
Central Richmond               2013.000000           394.422399   
Central Sunset                 2013.000000           423.687928   
Clarendon Heights              2012.000000           487.244886   
Corona Heights                 2012.400000           587.539067   
Cow Hollow                     2013.000000           665.964042   
Croker Amazon                  2012.833333           303.004184   
Diamond Heights                2011.500000           434.991739   
Downtown                       2013.000000           391.434378   
Duboce Triangle                2013.000000           502.599156   
Eureka Valley/Dolores Heights  2013.000000           642.248671   
Excelsior                      2013.333333           388.765927   
Financial District North       2013.000000           391.362533   
Financial District South       2012.400000           455.836212   
Forest Knolls                  2011.000000           321.840837   
Glen Park                      2013.166667           623.826178   
Golden Gate Heights            2012.600000           639.393557   
Haight Ashbury                 2013.000000           449.544762   
Hayes Valley                   2013.000000           355.932828   
Hunters Point                  2012.500000           170.624920   
Ingleside                      2012.500000           367.895144   
Ingleside Heights              2013.500000           384.797928   
Inner Mission                  2013.000000           397.673715   
Inner Parkside                 2013.750000           519.385604   
Inner Richmond                 2013.000000           378.594314   
Inner Sunset                   2013.000000           413.667515   
Jordan Park/Laurel Heights     2013.000000           529.384185   
Lake --The Presidio            2012.500000           409.954352   
Lone Mountain                  2012.500000           478.228553   
Lower Pacific Heights          2013.000000           539.237969   
Marina                         2013.000000           583.749269   
Merced Heights                 2014.000000           788.844818   
Midtown Terrace                2012.500000           550.510348   
Miraloma Park                  2011.750000           779.810842   
Mission Bay                    2012.666667           566.192214   
Mission Dolores                2012.500000           418.915623   
Mission Terrace                2013.600000           523.466201   
Nob Hill                       2013.000000           458.204094   
Noe Valley                     2013.000000           542.442913   
North Beach                    2013.400000           411.646876   
North Waterfront               2013.000000           498.269755   
Oceanview                      2012.333333           330.097385   
Outer Mission                  2013.500000           242.370952   
Outer Parkside                 2013.000000           485.027013   
Outer Richmond                 2013.000000           473.900773   
Outer Sunset                   2013.000000           394.219032   
Pacific Heights                2013.000000           689.555817   
Park North                     2013.000000           373.732856   
Parkside                       2012.500000           336.172661   
Parnassus/Ashbury Heights      2013.000000           632.740454   
Portola                        2012.000000           327.113925   
Potrero Hill                   2013.000000           662.013613   
Presidio Heights               2013.000000           675.350212   
Russian Hill      

### Step 3: Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. 

The second cell cleans the data and sets the “Neighborhood” column. 

Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

In [101]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())


Lat         Lon         year  sale_price_sqr_foot  \
Alamo Square     37.791012 -122.402100  2013.000000           366.020712   
Anza Vista       37.779598 -122.443451  2013.333333           373.382198   
Bayview          37.734670 -122.401060  2012.000000           204.588623   
Bayview Heights  37.728740 -122.410980  2015.000000           590.792839   
Bernal Heights   37.728630 -122.443050          NaN                  NaN   

                 housing_units   gross_rent  
Alamo Square          378401.0  2817.285714  
Anza Vista            379050.0  3031.833333  
Bayview               376454.0  2318.400000  
Bayview Heights       382295.0  3739.000000  
Bernal Heights             NaN          NaN

Lat        Lon    year  sale_price_sqr_foot  \
Yerba Buena      37.79298 -122.39636  2012.5           576.709848   
Bernal Heights        NaN        NaN  2013.5           576.746488   
Downtown              NaN        NaN  2013.0           391.434378   
Ingleside             NaN        NaN  2012.5           367.895144   
Outer Richmond        NaN        NaN  2013.0           473.900773   

                 housing_units   gross_rent  
Yerba Buena           377427.5  2555.166667  
Bernal Heights        379374.5  3080.333333  
Downtown              378401.0  2817.285714  
Ingleside             377427.5  2509.000000  
Outer Richmond        378401.0  2817.285714

In [104]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Lat         Lon         year  sale_price_sqr_foot  \
Alamo Square      37.791012 -122.402100  2013.000000           366.020712   
Anza Vista        37.779598 -122.443451  2013.333333           373.382198   
Bayview           37.734670 -122.401060  2012.000000           204.588623   
Bayview Heights   37.728740 -122.410980  2015.000000           590.792839   
Buena Vista Park  37.768160 -122.439330  2012.833333           452.680591   

                  housing_units   gross_rent  
Alamo Square           378401.0  2817.285714  
Anza Vista             379050.0  3031.833333  
Bayview                376454.0  2318.400000  
Bayview Heights        382295.0  3739.000000  
Buena Vista Park       378076.5  2698.833333

Lat         Lon     year  sale_price_sqr_foot  \
West Portal         37.74026 -122.463880  2012.25           498.488485   
Western Addition    37.79298 -122.435790  2012.50           307.562201   
Westwood Highlands  37.73470 -122.456854  2012.00           533.703935   
Westwood Park       37.73415 -122.457000  2015.00           687.087575   
Yerba Buena         37.79298 -122.396360  2012.50           576.709848   

                    housing_units   gross_rent  
West Portal             376940.75  2515.500000  
Western Addition        377427.50  2555.166667  
Westwood Highlands      376454.00  2250.500000  
Westwood Park           382295.00  3959.000000  
Yerba Buena             377427.50  2555.166667

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [106]:
# Create a plot to analyze neighborhood info
##import hvplot.pandas


all_neighborhoods_plot=all_neighborhoods_df.hvplot.points(
    'Lon',
    'Lat',
    geo=True,
    size='sale_price_sqr_foot',
    color='gross_rent',
    title='Sale Price & Rent Depending on Location',
    tiles = 'OSM',
    frame_width=700,
    frame_height=500
)

all_neighborhoods_plot

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot)

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** The neighborhood with the highest gross rent is Westwood Park, The neighborhood with the highest sale price per square foot is the Union Square District.


## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** # Gross rental income growth increased steadily whereas sales prices did not.

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** # I would recommend this strategy because while sales price varies, gross rent steadily increases.